### DISCUSSION TITLE
How to decompile API specific resources?

### DISCUSSION THREAD

**Discussion by sorgelig** [iBotPeaches/Apktool#3420] [Q&A]
### Information
1. **Apktool Version 2.9.0**
2. **Operating System Windows**
3. **APK From? ROM**
4. **Java Version JDK21**

I try to re-compile MiuiSystemUI.apk
Correct framework files are imported.

in original apk there are many API specific resources for example:
```
drawable-anydpi
drawable-anydpi-v21
drawable-anydpi-v23
drawable-anydpi-v24
drawable-anydpi-v26
drawable-hdpi
drawable-hdpi-v4
drawable-hdpi-v21
```

after decompiling with simple `apktool d` command, in res folder i see only:
```
drawable-anydpi
drawable-hdpi
```

so, there is no `-vXX` resources in res folder.
Obviously re-compiled MiuiSystemUI.apk doesn't work.

I've tried to google, but couldn't find any issues or guides related to this problem.
How to fix it?

**Comment by iBotPeaches** [comment]
There is no real fix for what you are describing - only workarounds. You can read the ROADMAP item here: https://github.com/iBotPeaches/Apktool/blob/master/ROADMAP.md#implicit-qualifiers-cleanup

Or an older response when I dug into this last here: https://github.com/iBotPeaches/Apktool/issues/1761#issuecomment-379309998

**Reply by sorgelig** [reply]
is it possible to simply have a command line switch to decompile all resource folders as-is and then compile them back as-is?
You already have versioning like anydpi, xdpi, xxdpi.. Why not allow -vXX in the folder names?

**Reply by iBotPeaches** [reply]
We do allow `-vXX` in the folder names - when a qualifier in the config detects it needed. You can see that here - https://github.com/iBotPeaches/Apktool/blob/master/brut.apktool/apktool-lib/src/main/java/brut/androlib/res/data/ResConfigFlags.java#L438-L460

Its not an option of introducing a command line switch - it doesn't work that way with how Apktool operates. You are looking at files and expecting an as-is output. It doesn't work that way. Apktool works off the resource table - its not crawling the file system with exception of unknown files.

**Reply by sorgelig** [reply]
Is it possible to decompile as it's do now, but at the end go through res folder in zip and decompile remaining resources? When it will be compiled back, you will do the same way to compile resources as it's already done and at the end compile the rest folders. I think with this way additional resources won't affect the work but if system will need API specific resources they will be available too.

**Reply by iBotPeaches** [reply]
You could change this line and remove the natural version requirement check - https://github.com/iBotPeaches/Apktool/blob/master/brut.apktool/apktool-lib/src/main/java/brut/androlib/res/data/ResConfigFlags.java#L428

This would write out the sdk version if its greater than 0, which may lead to other issues, but could glean some research from your investigation.
